src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -3,127 +3,166 @@ package org.rust.cargo.project.workspace.impl
      * the actual Library update happens on the event dispatch thread.
      */
-    override fun scheduleUpdate(toolchain: RustToolchain) {
+    override fun requestUpdateUsing(toolchain: RustToolchain, immediately: Boolean) {
         val contentRoot = module.cargoProjectRoot ?: return
 
-        val task = UpdateTask(toolchain, contentRoot.path, showFeedback = false)
-        alarm.cancelAllRequests()
-        alarm.addRequest({ task.queue() }, DELAY_MILLIS)
-    }
+        val delay = if (ApplicationManager.getApplication().isUnitTestMode) 0 else DELAY
 
-    override fun updateNow(toolchain: RustToolchain) {
-        val contentRoot = module.cargoProjectRoot ?: return
+        val task = UpdateTask(toolchain, contentRoot.path)
 
-        val task = UpdateTask(toolchain, contentRoot.path, showFeedback = true)
         alarm.cancelAllRequests()
-        task.queue()
-    }
 
-    @TestOnly
-    fun flushUpdates() {
-        alarm.flush()
+        if (immediately)
+            task.queue()
+        else
+            alarm.addRequest({ task.queue() }, delay, ModalityState.any())
     }
 
-    override fun loadState(state: CargoProjectState?) {
-        cargoProjectState = state ?: CargoProjectState()
-    }
+    /**
+     * Delivers latest cached project-description instance
+     *
+     * NOTA BENE: [CargoProjectWorkspace] is rather low-level abstraction around, `Cargo.toml`-backed projects
+     *            mapping underpinning state of the `Cargo.toml` workspace _transparently_, i.e. it doesn't provide
+     *            any facade atop of the latter insulating itself from inconsistencies in the underlying layer. For
+     *            example, [CargoProjectWorkspace] wouldn't be able to provide a valid [CargoProjectDescription] instance
+     *            until the `Cargo.toml` becomes sound
+     */
+    override val projectDescription: CargoProjectDescription?
+        get() = lock (lock) { cached }
 
-    override fun getState(): CargoProjectState = cargoProjectState
+    private fun notifyCargoProjectUpdate(r: UpdateResult) {
+        ApplicationManager.getApplication().assertIsDispatchThread()
 
-    override fun before(events: MutableList<out VFileEvent>) {
+        if (!module.isDisposed)
+            messageBus
+                .syncPublisher(CargoProjectWorkspaceListener.Topics.UPDATES)
+                .onWorkspaceUpdateCompleted(r)
     }
 
-    override fun after(events: MutableList<out VFileEvent>) {
-        if (!module.project.rustSettings.autoUpdateEnabled) return
-        val toolchain = module.toolchain ?: return
+    /**
+     * Subscribes given listener to the supplied topic on the [CargoProjectWorkspace]'s private
+     * message-bus
+     */
+    override fun <L> subscribeTo(t: Topic<L>, listener: L, disposer: Disposable?) {
+        val conn =
+            messageBus
+                .connect()
+                .apply {
+                    this.subscribe(t, listener!!)
```

Comment:
This `!!` can be avoided: just add `<L: Any>` bound to the `subscribeTo` function (the default is `L: Any?`). Also, I'd rather ban `!!` in non-test code :)